### DISCUSSION TITLE
Add 'with=' expression to simplify replacing record with a modified record

### DISCUSSION THREAD

**Discussion by Jarek300** [dotnet/csharplang#4298] [Language Ideas]
'with expression' is good for simple objects:

```
var p1 = new NamedPoint("A", 0, 0);
p1 = p1 with { Name = "B", X = 5 };
```
For complex objects 'with expression' causes large repetitions:

```
TrendSeries.Chart.LineWidth = TrendSeries.Chart.LineWidth with { Raw = value }; 
```

It would be much simpler to write

```
TrendSeries.Chart.LineWidth with= { Raw = value }; 
```

**Comment by alrz** [comment]
I like this. especially if it allows to modify non-records (eg ef entities)

**Comment by acaly** [comment]
<s>Dup #162. By the way, syntax like this has been discussed a lot in this repo and roslyn repo.</s>

Sorry for misunderstanding the question.

I think you are talking about the VB-style `with`, which has also been discussed a lot. For example: #2355, #2580.

**Comment by Jarek300** [comment]
No, it has nothing to do with VB-style (or earlier Pascal style) 'with'.

C#9 immutable records are only the first step. C#9 immutable records are good for vert simple cases but rather inconvenient in cases like above. Cases where object contains object that contains immutable record. Or immutable record that contains immutable record.

**Reply by AartBluestoke** [reply]
I should note that records are not guaranteed immutable, only defaulted that way... (your point is still correct though)

**Comment by JanTamis** [comment]
Wont this remove the immutable model that was created around records? why use records in this senario?

**Reply by spydacarnage** [reply]
It won't remove the immutability any more than the standard syntax - records are immutable only in that their data doesn't change, not the reference they point to, in the same way that strings are immutable - you can't change the string, but you can reassign a string variable.

Using the original example:
```csharp
var p1 = new NamedPoint("A", 0, 0);
p1 = p1 with { Name = "B", X = 5 };
p1 with= { Name = "B", X = 5 };  // the proposal is to make this synonymous with the line above
```

In saying that, I think it goes against the _spirit_ of a record, and I don't like it at all...

**Reply by AartBluestoke** [reply]
It gives you thread safety. You would be guaranteed to either see a fully formed original object or a fully formed updated object. No possibility of seeing an update with only name applied, and no locking needed.

**Comment by Jarek300** [comment]
Model is not immutable. Model is using immutable types. C# from version 1 has immutable types: string, DataTime, TimeSpan. C# 9 allows to easily create own immutable types thanks to records. 

Or I thought so.

In practise using immutable records starts to be unwieldy. Expression (or operator?) 'with=' could alleviate the problem.

**Reply by spydacarnage** [reply]
I'm confused.  If you want to be able to change a property of a object like that, make it `set`-able.
```csharp
TrendSeries.Chart.LineWidth.Raw = value; 
//or
TrendSeries.Chart.LineWidth.SetRaw(value);
```
It feels like you're trying to shoehorn records into a scenario which they're not really designed for.

**Comment by spydacarnage** [comment]
When you want to change more than property at the same time...  Yeah, okay
- I'll give you that one :)
